File: app/routers/synonyms.py
"""
同义词管理路由

Author: lvdaxianerplus
Date: 2026-06-05
"""

from pathlib import Path

from fastapi import APIRouter, HTTPException, Query, status

from app.config import Config
from app.models.knowledge_base_schemas import SynonymGroupCreateRequest
from app.models.knowledge_base_schemas import SynonymGroupUpdateRequest
from app.models.schemas import APIResponse
from app.services.knowledge_base_repository import KnowledgeBaseRepository


router = APIRouter(prefix="/api/v1/synonyms", tags=["Synonyms"])
KNOWLEDGE_BASE_DB_PATH = Config.KNOWLEDGE_BASE_DB_PATH or str(Path("data") / "knowledge_base.sqlite")


@router.get("", response_model=APIResponse)
async def list_synonym_groups(knowledge_base_id: str | None = Query(default=None)):
    """
    列出同义词组

    @param knowledge_base_id - 可选知识库 ID
    @returns 同义词组列表
    """
    data = _get_repository().list_synonym_groups(knowledge_base_id=knowledge_base_id)
    return APIResponse(code=200, message="success", data=data)


@router.post("", response_model=APIResponse)
async def create_synonym_group(request: SynonymGroupCreateRequest):
    """
    创建同义词组

    @param request - 同义词组创建请求
    @returns 创建后的同义词组
    """
    data = _get_repository().create_synonym_group(
        knowledge_base_id=request.knowledge_base_id,
        canonical=request.canonical,
        terms=request.terms,
        owner_id=request.owner_id,
        enabled=request.enabled,
    )
    return APIResponse(code=200, message="success", data=data)


@router.patch("/{group_id}", response_model=APIResponse)
async def update_synonym_group(group_id: str, request: SynonymGroupUpdateRequest):
    """
    更新同义词组

    @param group_id - 同义词组 ID
    @param request - 同义词组更新请求
    @returns 更新后的同义词组
    """
    try:
        data = _get_repository().update_synonym_group(
            group_id,
            request.model_dump(exclude_none=True),
        )
        return APIResponse(code=200, message="success", data=data)
    except ValueError as exc:
        raise _not_found(str(exc)) from exc


@router.delete("/{group_id}", response_model=APIResponse)
async def delete_synonym_group(group_id: str):
    """
    删除同义词组

    @param group_id - 同义词组 ID
    @returns 删除的同义词组 ID
    """
    try:
        data = _get_repository().delete_synonym_group(group_id)
        return APIResponse(code=200, message="success", data=data)
    except ValueError as exc:
        raise _not_found(str(exc)) from exc


def _get_repository() -> KnowledgeBaseRepository:
    """构建知识库仓储。"""
    return KnowledgeBaseRepository(KNOWLEDGE_BASE_DB_PATH)


def _not_found(message: str) -> HTTPException:
    """构建 404 响应异常。"""
    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={"code": 404, "message": message},
    )

File: web/src/api/synonyms.ts
/**
 * Recall · 同义词组 API 客户端
 *
 * 封装 `/api/v1/synonyms` 系列接口。SynonymGroup 用于 query 归一化与缓存 key 归一化。
 *
 * @author lvdaxianerplus
 */
import { requestJson } from "./client";
import type { ApiResponse } from "./types";

/**
 * 同义词组视图模型。
 *
 * @author lvdaxianerplus
 */
export interface SynonymGroup {
  id: string;
  knowledge_base_id: string | null;
  canonical: string;
  terms: string[];
  owner_id: string;
  enabled: boolean;
  created_at?: string;
  updated_at?: string;
}

/**
 * 同义词组写入 payload（创建 / 更新共用）。
 *
 * @author lvdaxianerplus
 */
export interface SynonymGroupPayload {
  knowledge_base_id?: string | null;
  canonical?: string;
  terms?: string[];
  owner_id?: string;
  enabled?: boolean;
}

/**
 * 列出同义词组。
 *
 * @param params.knowledge_base_id 可选按 KB 过滤
 * @returns 同义词组列表
 * @author lvdaxianerplus
 */
export async function listSynonymGroups(params?: { knowledge_base_id?: string }): Promise<SynonymGroup[]> {
  // 按 KB 过滤时把 id 编码后拼到 query string
  const query = params?.knowledge_base_id ? `?knowledge_base_id=${encodeURIComponent(params.knowledge_base_id)}` : "";
  const response = await requestJson<ApiResponse<SynonymGroup[]>>(`/api/v1/synonyms${query}`);
  return response.data;
}

/**
 * 创建同义词组。
 *
 * @param payload 同义词组配置
 * @returns 新建的同义词组
 * @author lvdaxianerplus
 */
export async function createSynonymGroup(payload: SynonymGroupPayload): Promise<SynonymGroup> {
  const response = await requestJson<ApiResponse<SynonymGroup>>("/api/v1/synonyms", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return response.data;
}

/**
 * 更新同义词组（PATCH）。
 *
 * @param id 同义词组 id
 * @param payload 字段 patch
 * @returns 更新后的同义词组
 * @author lvdaxianerplus
 */
export async function updateSynonymGroup(id: string, payload: SynonymGroupPayload): Promise<SynonymGroup> {
  const response = await requestJson<ApiResponse<SynonymGroup>>(`/api/v1/synonyms/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return response.data;
}

/**
 * 删除同义词组。
 *
 * @param id 同义词组 id
 * @returns 包含被删 id 的确认
 * @author lvdaxianerplus
 */
export async function deleteSynonymGroup(id: string): Promise<{ id: string }> {
  const response = await requestJson<ApiResponse<{ id: string }>>(`/api/v1/synonyms/${id}`, {
    method: "DELETE",
  });
  return response.data;
}
